### PR TITLE
Add database import/export feature

### PR DESCRIPTION
**Pull request by ZvonimirMedak**
## Summary

Added a database import/export feature which will replace the existing database file

### Type

- [X] **Feature**: This pull request introduces a new feature.
- [ ] **Bug fix**: This pull request fixes a bug.
- [ ] **Refactor**: This pull request refactors existing code.
- [ ] **Documentation**: This pull request updates documentation.
- [ ] **Other**: This pull request makes other changes.

#### Additional information

- [ ] This pull request introduces a **breaking change**.

### Description

The Database import export requires the UTType so Sentinel knows which files it can import. The database should be provided as a relative path to the documents URL of the App

## Checklist

- [X] I have performed a self-review of my own code.
- [X] I have tested my changes, including edge cases.
- X ] I have added necessary tests for the changes introduced (if applicable).
- [X] I have updated the documentation to reflect my changes (if applicable).

## Additional notes

I have the recordings of how it works, so please reach out to me, and I'll send you via DM since they have some project information.

### CODE REVIEW

File: Sentinel/Classes/DatabaseImportExport/DatabaseImportExportTool.swift [hunk 1]
```diff
@@ -0,0 +1,63 @@
...
+//  Sentinel
+//
+//  Created by Zvonimir Medak on 12.03.2025..
+//
+
+import SwiftUI
+import UniformTypeIdentifiers
+
+/// Tool which gives the ability to import or export the database file
+public struct DatabaseImportExportTool: Tool {
```
Comment (nikolamajcen) [inline Sentinel/Classes/DatabaseImportExport/DatabaseImportExportTool.swift:12]: I would make this `DatabaseTool`, and functions inside are import/export, but in future we can extend that.

File: README.md [hunk 2]
```diff
@@ -283,6 +283,10 @@ var colorChangeTool: Tool {
 
 *CrashDetectionTool* sets up listeners on system exceptions which will be logged into a local json file with the stack trace. The tool is available on iOS.
 
+#### Database Import/Export tool
```
Comment (nikolamajcen) [inline README.md:286]: Based on the proposed change, this name may change as well - if you'll agree on that. If not, it can stay like that.

File: Sentinel/Classes/DatabaseImportExport/DatabaseImportExportView.swift [hunk 3]
```diff
@@ -0,0 +1,122 @@
...
+
+struct DatabaseImportExportView: View {
+
+    @ObservedObject var viewModel: DatabaseImportExportViewModel
+
+    @State private var showPicker = false
+    @State private var showShare = false
+
+    var body: some View {
+        VStack(spacing: 30) {
```
Comment (nikolamajcen) [inline Sentinel/Classes/DatabaseImportExport/DatabaseImportExportView.swift:19]: I would recommend using 4, 8, 12, 16, 24, 32, if that's fine with you.

File: Sentinel/Classes/DatabaseImportExport/DatabaseImportExportView.swift [hunk 4]
```diff
@@ -0,0 +1,122 @@
...
+            Button(action: { showPicker = true }) {
+                Text("Import Database")
```
Comment (nikolamajcen) [inline Sentinel/Classes/DatabaseImportExport/DatabaseImportExportView.swift:21]:
```suggestion
                Text("Import database")
```

File: Sentinel/Classes/DatabaseImportExport/DatabaseImportExportView.swift [hunk 5]
```diff
@@ -0,0 +1,122 @@
...
+            }
+
+            Button(action: { viewModel.exportDatabase() }) {
+                Text("Export Database")
```
Comment (nikolamajcen) [inline Sentinel/Classes/DatabaseImportExport/DatabaseImportExportView.swift:25]:
```suggestion
                Text("Export database")
```

File: Sentinel/Classes/DatabaseImportExport/DatabaseImportExportView.swift [hunk 6]
```diff
@@ -0,0 +1,122 @@
...
+            )
+            #else
+            .sheet(isPresented: $showShare) {
+                ActivityViewController(
+                    activityItems: [viewModel.selectedURL as Any],
+                    applicationActivities: nil,
+                    didFinish: { viewModel.selectedURL = nil }
+                )
+            }
+            #endif
```
Comment (nikolamajcen) [inline Sentinel/Classes/DatabaseImportExport/DatabaseImportExportView.swift:51]: I believe this can be wrapped in one single component, and calling side would not be aware of the platform.

File: Sentinel/Classes/DatabaseImportExport/Internal/DatabaseFileManager.swift [hunk 7]
```diff
@@ -0,0 +1,95 @@
...
+    }()
+
+    // MARK: - Init
+
+    init(fileManager: FileManager = .default, databaseFilePath: String) {
+        self.fileManager = fileManager
+        self.databaseFilePath = databaseFilePath
+    }
+
+    func importDatabase(from url: URL) throws {
```
Comment (nikolamajcen) [inline Sentinel/Classes/DatabaseImportExport/Internal/DatabaseFileManager.swift:46]: Based on import/export functionality, do we expect the concrete file here or the `.zip` as well? I would like to handle both cases, because, if we are exporting zip, we could expect to get zip as an import as well.

File: Sentinel/Classes/DatabaseImportExport/Internal/DatabaseFileManager.swift [hunk 8]
```diff
@@ -0,0 +1,95 @@
...
+        guard
+            let databaseURL,
+            url.startAccessingSecurityScopedResource()
+        else { return }
+        deleteDirectoryIfExists(atPath: databaseURL.path)
+
+        _ = try fileManager.copyItem(at: url, to: databaseURL)
+    }
+
+    func exportDatabase() -> URL? {
```
Comment (nikolamajcen) [inline Sentinel/Classes/DatabaseImportExport/Internal/DatabaseFileManager.swift:57]: Regarding the export, should we provide option to export the database itself (not an archive) and the archived file? I'm just asking due to the consistency, feel free to argument me on those two comments 😄👍

File: Sentinel/Classes/DatabaseImportExport/Internal/ZipHandler.swift [hunk 9]
```diff
@@ -0,0 +1,51 @@
...
+import Foundation
+
+enum CreateZipError: Error {
+    case urlNotADirectory(URL)
+    case failedToCreateZIP(Error)
+}
+
+enum ZipHandler {
+
+    static func createZip(zipFinalURL: URL, fromDirectory directoryURL: URL, databaseURL: URL) throws -> URL {
```
Comment (nikolamajcen) [inline Sentinel/Classes/DatabaseImportExport/Internal/ZipHandler.swift:17]: `zipFinalURL` -> `zipDestinationURL`

File: Sentinel/Classes/DatabaseImportExport/Internal/ZipHandler.swift [hunk 10]
```diff
@@ -0,0 +1,51 @@
...
+//
+
+import Foundation
+
+enum CreateZipError: Error {
+    case urlNotADirectory(URL)
+    case failedToCreateZIP(Error)
+}
+
+enum ZipHandler {
```
Comment (nikolamajcen) [inline Sentinel/Classes/DatabaseImportExport/Internal/ZipHandler.swift:15]: Maybe go with `Archive` instead of `Zip`? I'm fine with both, but a bit creative today 😄

**Review by nikolamajcen — CHANGES_REQUESTED** [review]
Really nice 👍 I've added some comments, but overall this is great. Once we close open questions and do changes (if needed), I'll approve this one.

File: Sentinel/Classes/DatabaseImportExport/DatabaseImportExportView.swift (hunk 6 above)
Comment (ZvonimirMedak) [reply Sentinel/Classes/DatabaseImportExport/DatabaseImportExportView.swift:51]: nice one 👌

File: Sentinel/Classes/DatabaseImportExport/Internal/DatabaseFileManager.swift (hunk 7 above)
Comment (ZvonimirMedak) [reply Sentinel/Classes/DatabaseImportExport/Internal/DatabaseFileManager.swift:46]: I have added the export as zip, and file but the unzip is complicated. Atleast, from what I have found online, there's no easy was to unzip a file using built-in functions of Swift. Only the archiving is kinda easy with the NSFileCoordinator since the forUploading option zips it.
And I wouldn't want to add a third party too just so we can accept a zip since I believe mostly people will just export the file or zip (depending on the size), and import the file.

Do you have an example or an idea how we can achieve the unarchiving using built-in functions?

Comment (nikolamajcen) [reply Sentinel/Classes/DatabaseImportExport/Internal/DatabaseFileManager.swift:46]: I agree with you 👍

File: Sentinel.podspec [hunk 11]
```diff
@@ -62,6 +62,11 @@ Pod::Spec.new do |s|
     sp.source_files = 'Sentinel/Classes/CrashDetection/**/*'
     sp.dependency 'Sentinel/Core'
   end
+
+  s.subspec 'Database' do |sp|
+    sp.source_files = 'Sentinel/Classes/DatabaseImportExport/**/*'
```
Comment (nikolamajcen) [inline Sentinel.podspec:67]: I would suggest changing path to use `DatabaseTool` instead of `DatabaseImportExport`.

**Review by nikolamajcen — APPROVED** [review]
Nice work!
